examples/firewall/firewall.c

```diff
@@ -0,0 +1,388 @@
+                                rules_init = 1;
+                                break;
+                        case 'b':
+                                RTE_LOG(INFO, APP,
+                                        "Debug mode enabled; printing the source IP addresses of each incoming packet as well as drop/forward status\n");
+                                debug = 1;
+                                break;
+                        case '?':
+                                usage(progname);
+                                if (optopt == 'p')
+                                        RTE_LOG(INFO, APP, "Option -%c requires an argument.\n", optopt);
+                                if (optopt == 'd')
+                                        RTE_LOG(INFO, APP, "Option -%c requires an argument.\n", optopt);
+                                if (optopt == 'f')
+                                        RTE_LOG(INFO, APP, "Option -%c requires an argument.\n", optopt);
+                                else if (isprint(optopt))
+                                        RTE_LOG(INFO, APP, "Unknown option `-%c'.\n", optopt);
+                                else
+                                        RTE_LOG(INFO, APP, "Unknown option character `\\x%x'.\n", optopt);
+                                return -1;
+                        default:
+                                usage(progname);
+                                return -1;
+                }
+        }
+
+        if (!dst_flag) {
+                RTE_LOG(INFO, APP, "Firewall NF requires a destination NF with the -d flag.\n");
+                return -1;
+        }
+        if (!debug) {
+                RTE_LOG(INFO, APP, "Running normal mode, use -b flag to enable debug mode\n");
+        }
+        if (!rules_init) {
+                RTE_LOG(INFO, APP, "Please specify a rules JSON file with -f FILE_NAME\n");
+                return -1;
+        }
+        return optind;
+}
+
+/*
+ * This function displays stats. It uses ANSI terminal codes to clear
+ * screen when called. It is called from a single non-master
+ * thread in the server process, when the process is run with more
+ * than one lcore enabled.
+ */
+static void
+do_stats_display(void) {
+        const char clr[] = {27, '[', '2', 'J', '\0'};
+        const char topLeft[] = {27, '[', '1', ';', '1', 'H', '\0'};
+
+        /* Clear screen and move to top left */
+        printf("%s%s", clr, topLeft);
+        printf("Packets Dropped: %lu\n", stats.pkt_drop);
+        printf("Packets Accepted: %lu\n", stats.pkt_accept);
+        printf("Packets Total: %lu", stats.pkt_total);
+
+        printf("\n\n");
+}
+
+static int
+packet_handler(struct rte_mbuf *pkt, struct onvm_pkt_meta *meta, __attribute__((unused)) struct onvm_nf_info *nf_info) {
+        struct ipv4_hdr *ipv4_hdr;
+        static uint32_t counter = 0;
+        uint32_t rule = 0;
+        uint32_t track_ip = 0;
+
+        if (onvm_pkt_is_ipv4(pkt)) {
```

Comment:
Instead of multiple levels of if/else branching try to check conditions and return on a negative outcome, this will make your code cleaner and easier to read. Check[ the loadbalncer for example](https://github.com/sdnfv/openNetVM-dev/blob/develop/examples/load_balancer/load_balancer.c#L491).

So here it would mean
```
if (onvm_pkt_is_ipv4(pkt)) {
// this should probably be a different field, something like stats.pkt_not_ipv4 or something
                stats.pkt_drop++;
                stats.pkt_total++;
                meta->action = ONVM_NF_ACTION_DROP;
                meta->destination = 0;
                return 0;
}

ret = rte_lpm_lookup(lpm_tbl, ipv4_hdr->src_addr, &rule);
if (ret < 0) {
                stats.pkt_drop++;
                stats.pkt_total++;
                meta->action = ONVM_NF_ACTION_DROP;
                meta->destination = 0;
                return 0;
}

Then you can go into your 
switch (rule) ...

```